### PR TITLE
add the architecture doc

### PR DESCRIPTION
**Pull request by think-in-universe**
### Description

Add the architecture doc for the issue https://github.com/protofire/solhint/issues/140

### Details

1. Project Structure
1. ANTLR Grammar with [solidity-antlr4](https://github.com/solidityj/solidity-antlr4)
1. How to Add a New Rule, A Plugin or A Shared Configuration

### CODE REVIEW

**Comment by think-in-universe** [comment]
@fvictorio sure. thanks. 😄

**Comment by think-in-universe** [comment]
@ChrisChinchilla any feedback or suggestions?

**Comment by think-in-universe** [comment]
Hi @fvictorio @ChrisChinchilla, any update for the document?

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+title:       Solhint Project Structure
+date:        2019-08-27 16:11:00 +0000
+author:      "@think-in-universe"
+description: Introduction of the structure of the project, how it works, and how to add new rules, etc.
+---
+
+
+# Architecture
+
+This document is written for the developers who want to understand the structure and mechanism of this project, and how to contribute.
```
Comment (ChrisChinchilla) [inline docs/architecture.md]:
```suggestion
This document is for developers who want to understand the structure and mechanisms of this project, and how to contribute.
```

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+
+## Project Structure
+
+Below describes the major structure of the project.
```
Comment (ChrisChinchilla) [inline docs/architecture.md]:
```suggestion
The file structure below describes the major structure of the project.
```

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+│   ├── reporter.js                      # module for reporting results
+│   ├── rules                            # source for solhint rules
+│   └── tree-listener.js                 # used for register all loaded rules
+├── scripts                              # script for generating grammar and rule docs
+├── solhint.js                           # solhint command line
+├── solidity-antlr4                      # git submodule for solidity-antlr4
+└── test                                 # tests
+```
+
+## How it Works
```
Comment (ChrisChinchilla) [inline docs/architecture.md]:
```suggestion
## How solhint Works
```

File: docs/architecture.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+### Solidity Grammar with ANTLR
+
+Solhint depends on [ANTLR4](http://www.antlr.org/) to generate the Solidity parser, following the grammar description taken from
+[solidity-antlr4](https://github.com/solidityj/solidity-antlr4).
+
+To update the Solidity grammar, you'll need to update the Git submodule solidity-antlr4:
```
Comment (ChrisChinchilla) [inline docs/architecture.md]:
```suggestion
To update the Solidity grammar, you need to update the Git submodule solidity-antlr4:
```

File: docs/architecture.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+```sh
+git submodule update --remote # update to latest version
+# or checkout a specific commit or tag:
+# cd solidity-antlr4
+# git checkout abc123
+```
+
+Then execute `scripts/build-grammar.sh`. (Java Runtime Environment 7, or later, is required for running the script.)
```
Comment (ChrisChinchilla) [inline docs/architecture.md]:
```suggestion
Then run `scripts/build-grammar.sh`. (Java Runtime Environment 7, or later, is required for running the script.)
```

File: docs/architecture.md [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+### How to Add A New Rule
+
+The Solhint rules in `lib/rules` contains the different aspects of lint requirements, such as naming, best practices, security, etc.
```
Comment (ChrisChinchilla) [inline docs/architecture.md]:
```suggestion
The Solhint rules in `lib/rules` contains the different lint requirements, such as naming, best practices, security, etc.
```

File: docs/architecture.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+
+The rules are implemented with a visitor pattern. You can extend the `BaseChecker` class with the `ruleId` and `meta` fields to define a rule, and implement methods that are called when a node in the AST is entered or exited. The constructor accepts a reporter and a config, and `ruleId` field is present in the object. This `ruleId` is the one that will be used to activate and configure the rule.
```
Comment (ChrisChinchilla) [inline docs/architecture.md]:
```suggestion
The rules are implemented with a [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern). You can extend the `BaseChecker` class with the `ruleId` and `meta` fields to define a rule, and implement methods that are called when a node in the AST is entered or exited. The constructor accepts a reporter and a config, and `ruleId` field is present in the object. This `ruleId` is the one that will be used to activate and configure the rule.
```

File: docs/architecture.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    this.blockValidator.validateBlock(ctx)
+  }
+
+  enterImportDirective(ctx) {
+    this.blockValidator.validateBlock(ctx)
+  }
+
+```
+
+The developers of the new rules need to have basic understanding about the concepts and structure of the AST, and execute the proper logic when certain nodes in the AST are visited.
```
Comment (ChrisChinchilla) [inline docs/architecture.md]:
```suggestion
Developers of new rules need to have a basic understanding about the concepts and structure of the AST, and execute the proper logic when certain nodes in the AST are visited.
```

**Comment by ChrisChinchilla** [comment]
@think-in-universe Couple of small things and we're good to go 👍

**Comment by think-in-universe** [comment]
@ChrisChinchilla cool. Thanks for the suggestions and I have updated the doc.

**Comment by ChrisChinchilla** [comment]
Merge away @fvictorio ! Then we'll pay the bounty out asap too.

**Comment by fvictorio** [comment]
Thanks @think-in-universe!

**Comment by think-in-universe** [comment]
thank you guys @fvictorio @ChrisChinchilla

**Pull request merged**